apps/condo/domains/banking/tasks/generateReports.js

```diff
@@ -0,0 +1,270 @@
+        }
+        for (let i = 0; i < transactions.length; i++) {
+            const currentTransaction = transactions[i]
+            if (Object.keys(costItemData.data).includes(currentTransaction.costItem.id)) {
+                costItemData.data[currentTransaction.costItem.id].amount += Number(currentTransaction.amount)
+            } else {
+                costItemData.data[currentTransaction.costItem.id] = {
+                    amount: Number(currentTransaction.amount),
+                    name: currentTransaction.costItem.name,
+                    //isOutcome: currentTransaction.isOutcome,
+                    category: {
+                        id: currentTransaction.costItem.category.id,
+                        name: currentTransaction.costItem.category.name,
+                    },
+                }
+            }
+
+            await updateTaskProgress({
+                task,
+                lastProgress,
+                context,
+                progress: allTransactions.length / transactions.length / (i + 1) * 100,
+            })
+            lastProgress = Date.now()
+        }
+        costItemsDataSortedByDate.push(costItemData)
+    }
+
+    const categoryGroupsSortedByMonth = await categoryGroupsBuilder(
+        costItemsDataSortedByDate,
+        {
+            task,
+            lastProgress,
+            context,
+        })
+    lastProgress = Date.now()
+
+    for (let { date, data } of categoryGroupsSortedByMonth) {
+        let index = 0
+        const reports = await BankAccountReport.getAll(context, {
+            period: date,
+            deletedAt: null,
+            organization: { id: organization.id },
+        }, { sortBy: ['createdAt_ASC'] })
+        const payload = {
+            data,
+            version: 1,
+            period: date,
+            account: { connect: { id: bankAccountId } },
+            organization: { connect: { id: organization.id } },
+            totalIncome: monthTurnovers[index].totalIncome,
+            totalOutcome: monthTurnovers[index].totalOutcome,
+            template: EXPENSES_GROUPED_BY_CATEGORY_AND_COST_ITEM,
+            amount: get(bankAccount, 'meta.amount'),
+            amountAt: get(bankAccount, 'meta.amountAt'),
+            ...DV_SENDER,
+        }
+        if (reports.length > 0 && isEqual(reports[reports.length - 1].data, data[index])) {
+            if (date !== monthTurnovers[index].date) throw new Error('Date from categoryGroups not equal date from monthTurnovers')
+            await BankAccountReport.create(context, {
+                ...payload,
+                version: ++reports[reports.length - 1].version,
+            })
+            index++
+        } else {
+            await BankAccountReport.create(context, payload)
+        }
+        await updateTaskProgress({
+            task,
+            lastProgress,
+            context,
+            progress: categoryGroupsSortedByMonth.length / (index + 1) * 100,
+        })
```

Comment:
Should we sleep at the end of iteration?

Reply:
I don't know, it's just js code and a couple of progress updates, they are limited by the interval, there is no load here, I see no reason to sleep